### PR TITLE
Feature/586 block diagonal direct

### LINKED ISSUE #586: Adapt `DirectInfluence` to use block-approximation
Extend `DirectInfluence` to handle:

1. Block-diagonal approximations
2. Gauss-Newton approximation

### PR DESCRIPTION
**Pull request by schroedk**
### Description

This PR closes #586 

### Changes

- Add functionality for block-diagonal and Gauss-Newton approximation to `DirectInfluence`

### Checklist

- [x] Wrote Unit tests (if necessary)
- [x] Updated Documentation (if necessary)
- [x] Updated Changelog
- [x] If notebooks were added/changed, added boilerplate cells are tagged with `"tags": ["hide"]` or `"tags": ["hide-input"]`

### CODE REVIEW

**Comment by schroedk** [comment]
Blocked by #582

File: docs/influence/index.md [hunk 1]
```diff
@@ -249,6 +249,87 @@ not to corrupt the outcome too much, the parameter $\lambda$ should be as small
...
+))
+# custom block-diagonal structure
+infl_model = DirectInfluence(model, loss,
+                             regularization=0.1,
+                             block_structure=block_structure)
+infl_model = infl_model.fit(training_data_loader)
+```
+If you would like to apply a block-specific regularization, you can provide a
+dictionary with the block names as keys and the regularization values as values.
+In this case, the specification must be complete, i.e. every block must have
```
Comment (janosg) [inline docs/influence/index.md:285]: We could also fill everything for which no regularization was provided with a default value (probably 0)

File: src/pydvl/influence/torch/functional.py [hunk 2]
```diff
@@ -424,13 +437,16 @@ def hessian(
         def flat_input_batch_loss_function(
             p: torch.Tensor, t_x: torch.Tensor, t_y: torch.Tensor
         ):
-            return blf(align_with_model(p, model), t_x, t_y)
+            return blf(align_structure(params, p), t_x, t_y)
```
Comment (janosg) [inline src/pydvl/influence/torch/functional.py:440]: Took me a while to understand that `blf` means `batch_loss_function`. I think `batch_loss` would be much clearer than `blf` and not too long (unless this is again a strong pytorch convention or so)

File: src/pydvl/influence/torch/functional.py [hunk 3]
```diff
@@ -447,6 +463,57 @@ def flat_input_empirical_loss(p: torch.Tensor):
...
+
+    Returns:
+        The Gauss-Newton matrix.
+    """
+
+    per_sample_grads = create_per_sample_gradient_function(model, loss)
+
+    params = restrict_to
+    if params is None:
+        params = {k: p.detach() for k, p in model.named_parameters() if p.requires_grad}
```
Comment (janosg) [inline src/pydvl/influence/torch/functional.py:496]: Dict comprehensions like this appear very often. Should we have a function `get_detached_params(model)` instead?

File: src/pydvl/influence/torch/influence_function_model.py [hunk 4]
```diff
@@ -351,109 +353,115 @@ def to(self, device: torch.device):
...
+        ```
+        Accordingly, if you choose a layer-wise or parameter-wise structure
+        (by providing `BlockMode.LAYER_WISE` or `BlockMode.PARAMETER_WISE` for
+        `block_structure`) the keys must be the layer names or parameter names,
+        respectively.
+
+        You can retrieve the block-wise influence information from the methods
+        with suffix `_by_block`. By default, `block_structure` is set to
+        `BlockMode.FULL` and in this case these methods will return a dictionary
+        with the empty string being the only key.
```
Comment (janosg) [inline src/pydvl/influence/torch/influence_function_model.py:398]: This text is very similar to the description in `docs/influence/index.md`. Could we link to that place instead of repeating to make maintenance easier?

File: src/pydvl/influence/torch/influence_function_model.py [hunk 5]
```diff
@@ -351,109 +353,115 @@ def to(self, device: torch.device):
...
+        block_params: Dict[str, torch.nn.Parameter],
+        data: DataLoader,
+        regularization: Optional[float],
+    ) -> TorchOperatorGradientComposition:
+        gp = TorchGradientProvider(self.model, self.loss, restrict_to=block_params)
 
-        """
-        return super().influences(x_test, y_test, x, y, mode=mode)
+        if self.second_order_mode is SecondOrderMode.GAUSS_NEWTON:
+            mat = gauss_newton(self.model, self.loss, data, restrict_to=block_params)
```
Comment (janosg) [inline src/pydvl/influence/torch/influence_function_model.py:455]: More a question than a comment: Is it efficient to materialize the Gauss Newton matrix and then use a direct solver? I thought it would be more efficient (and numerically robust) to build up the solutions from the factors of the Gauss Newton matrix (e.g. something like QR decomposition on the factors, triangular solve on R and then multiply).

File: docs/influence/index.md (hunk 1 above)
Comment (schroedk) [reply docs/influence/index.md:285]: This is the actual behavior, but the documentation is wrong, as it describes a special case.

File: src/pydvl/influence/torch/functional.py (hunk 3 above)
Comment (schroedk) [reply src/pydvl/influence/torch/functional.py:496]: Agree, will check

File: src/pydvl/influence/torch/influence_function_model.py (hunk 4 above)
Comment (schroedk) [reply src/pydvl/influence/torch/influence_function_model.py:398]: Agree, will check

File: src/pydvl/influence/torch/influence_function_model.py (hunk 5 above)
Comment (janosg) [reply src/pydvl/influence/torch/influence_function_model.py:455]: I created Issue #594 for this

File: src/pydvl/influence/torch/functional.py (hunk 2 above)
Comment (schroedk) [reply src/pydvl/influence/torch/functional.py]: Renamed to `batch_loss`, done.

File: src/pydvl/influence/torch/functional.py (hunk 3 above)
Comment (schroedk) [reply src/pydvl/influence/torch/functional.py]: Add a util function `get_model_parameters`, done.

File: src/pydvl/influence/torch/influence_function_model.py (hunk 4 above)
Comment (schroedk) [reply src/pydvl/influence/torch/influence_function_model.py]: Linked to documentation, done.

**Review by janosg — APPROVED** [review]
